kernel/src/engine/arrow_expression.rs

```diff
@@ -245,12 +243,14 @@ fn evaluate_expression(
                 .iter()
                 .zip(schema.fields())
                 .map(|(array, input_field)| {
-                    let output_data_type =
-                        ensure_data_types(input_field.data_type(), array.data_type());
-                    // use a `map` here so rustc doesn't have to infer the error type
-                    output_data_type.map(|output_type| {
-                        ArrowField::new(input_field.name(), output_type, array.is_nullable())
-                    })
+                    ensure_data_types(input_field.data_type(), array.data_type())?;
+                    // need to help type inference a bit so it knows what the error type is
+                    let res: DeltaResult<ArrowField> = Ok(ArrowField::new(
```

Comment:
nit: I suspect this would also work:
```rust
.map(|(array, input_field)| -> DeltaResult<_> {
    ensure_data_types(...)?;
    Ok(ArrowFIeld::new(...));
}
```